### PR TITLE
refactor(main): Use repositories and rework tests

### PR DESCRIPTION
**Pull request by kjdelisle**
Changes:
- Add repositories
- Refactor to accommodate their use
- Rewrite majority of tests as unit tests
- Rewrite acceptance test as a basic suite of CRUD operations against the Todo API.

### CODE REVIEW

File: src/controllers/todo-item.controller.ts
Comment (jannyHou) [inline src/controllers/todo-item.controller.ts]: hmm...is it `replaceById` or `deleteById`?

File: src/controllers/todo-item.controller.ts
Comment (jannyHou) [inline src/controllers/todo-item.controller.ts]: you probably want to adjust the @param sequence accordingly as well, like what you've changed in `deleteTodoItem`

File: package.json
Comment (bajtos) [inline package.json]: FWIW, we are using `\"` intead of `'` in other places. IIIRC, `'` does not work on Windows, but I am not sure.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: IIUC, this test is depending on another test to create an existing `item`. This is a huge anti-pattern. One of the negative effects is that you cannot run the test `successfully deletes todos` on its own (e.g. via `it.skip`).

Please use test data builders as described in [Thinking in LoopBack](http://loopback.io/doc/en/lb4/Thinking-in-LoopBack.html#update-test-helpers-and-the-controller-use-real-model-and-repository) and [Testing your Application](http://loopback.io/doc/en/lb4/Testing-your-application.html#data-handling).

When I was writing that content, it did not come to my mind that we need to warn users to not write interdependent tests. I created a task to improve that part of our docs - see https://github.com/strongloop/loopback-next/issues/683

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: Implementation of helper/hook functions belongs to the bottom, see http://loopback.io/doc/en/contrib/style-guide.html#layout-of-test-files

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: I find it weird to have complex variables like `todo` and `payload` declared directly inside `describe` block. Shouldn't we create new instances for each test via `beforeEach`?

I think the setup can be also simplified - juggler models accept partial data in the constructor.

```ts
const todo = new Todo({id: 1, title: 'things to do today', isComplete: false});
const payload = new TodoItem({title: 'do a thing'});
```

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: Please add an empty line between each pair of tests to serve as a visual delimiter, to make it easier to see where one test finishes and another starts.

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: When the controller does not thrown any error, this assertion is skipped. Should you perhaps add another `throw new Error('createTodoItem should have thrown an error')` at the end of the `try` block?

The same comment applies to other similar tests below.

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: Have you considered using a test data builder instead, as described in http://loopback.io/doc/en/lb4/Testing-your-application.html#use-test-data-builders?

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: In my experience, this style where the stubs have strict argument matching produces test failures that are difficult to troubleshoot.

For example, if the controller invokes `replace` with id `42` instead of `item.id`, then `replace` returns undefined and sometime later you will see an error that `undefined` was supposed to equal to something else.

Another problem with this approach is that it breaks arrange-act-assert pattern - the assertion that `replace` should be invoked with `item.id` is implicitly hidden in the arrange part where we are setting up the stubbed behavior.

A better approach is to make stubs loose, let them always return the good value regardless of the input arguments. Then add assertions to verify that stubbed method were invoked as expected.

```ts
// arrange
replace.resolves(newItem);

// act
// controller.replaceTodoItem

// assert
sinon.assert.calledWith(replace, item.id as number, newItem);
```

This produces a more helpful test failure:

```
AssertError: expected stub to be called with arguments
1 2
// 1 is printed in green, 2 is printed in red
```

The good approach is described in http://loopback.io/doc/en/lb4/Testing-your-application.html#create-a-stub-repository, but the section does not say why other approaches are bad.  See https://github.com/strongloop/loopback-next/issues/684

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: I find these unit tests too similar to each other, the whole file looks like a big wall of text to me.

Could you please extract parts that are repeating into a shared test setup?

For example, I think that all `item`, `newItem`, `controller`, `update`, `findTodoItem`, `create`, etc. can be defined in a single `beforeEach` hook. 

It's ok that the hook configures more stuff than is used by each test, because creating extra stubs are easy. What's more important is to make the tests easy to read and understand, make it easy for the reader to figure out what is important in each test, what's the difference from other tests.

File: test/unit/controllers/todo-item.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo-item.controller.test.ts]: We don't keep commented-out blocks of code. Please uncomment and make the code work, or remove it completely.

File: test/unit/controllers/todo.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo.controller.test.ts]: Please apply all my comments for `todo-item.controller.test.ts` to this test file too.

File: src/controllers/todo-item.controller.ts
Comment (bajtos) [inline src/controllers/todo-item.controller.ts]: This pattern where each controller method calls `findById` first and then performs the operation, is creating a race condition. For example, what if another clients deletes the todoItem instance after `findById` was finished but before `updateById` below was called?

IMO, the controller methods should not be calling `findById` at all. It's the job of the database connector (or the juggler DAO layer) to return a meaningful error when we are attempting to edit or delete an object that does not exist. Only the connector is able to check for existence and modify/delete the model instance in an atomic way.

**Review by bajtos — CHANGES_REQUESTED** [review]
Nice progress! Let's improve this little bit more.

File: src/controllers/todo-item.controller.ts
Comment (kjdelisle) [reply src/controllers/todo-item.controller.ts]: Had a refactor snafu there that goofed up a ton of code. Thanks for the catch!

**Comment by kjdelisle** [comment]
@bajtos PTAL

The errors returned from legacy-juggler-bridge are very inconsistent IMO.
I wanted to handle the lack of certain outputs at the controller level because I don't think protocol-specific error codes should be the responsibility of the repository layer.

Even ignoring this, the return values for many of the id-based functions are very inconsistent and won't return errors when no object with the given id exists (which is a big problem).

I would've expected each of the id-based crud methods to return a 404 if there is no valid target, but the `updateById` function returns 200 OK (even though nothing is updated) and `deleteById` returns 400 Bad Request.

In any case, I'd like to land this PR if I can with just the happy-path test changes for now (to limit how long I'm stuck on this) and revisit this after we've had a chance to review the legacy-juggler APIs.

**Comment by bajtos** [comment]
> The errors returned from legacy-juggler-bridge are very inconsistent IMO.
I wanted to handle the lack of certain outputs at the controller level because I don't think protocol-specific error codes should be the responsibility of the repository layer.

I agree protocol-specific error codes should not be the responsibility of the repository layer. I think this is the legacy of our LoopBack 3.x design.

> Even ignoring this, the return values for many of the id-based functions are very inconsistent and won't return errors when no object with the given id exists (which is a big problem).
>
> I would've expected each of the id-based crud methods to return a 404 if there is no valid target, but the updateById function returns 200 OK (even though nothing is updated) and deleteById returns 400 Bad Request.

I agree this is a big problem, could you please open a new GH issue to keep track of this? Ideally with enough information that will allow us to reproduce those issues.

> In any case, I'd like to land this PR if I can with just the happy-path test changes for now (to limit how long I'm stuck on this) and revisit this after we've had a chance to review the legacy-juggler APIs.

Makes sense 👍 

And now I am going to review the patch itself :)

File: package.json
Comment (bajtos) [inline package.json]: FWIW, we are not (or should not be) applying prettier formatting to `package.json`, because `npm` will reformat the file back to `JSON.stringify` format whenever we change it, typically via `npm install`. Could you please revert these white-space only changes?

I am not sure if `package.json` is listed in `.prettierignore`, feel free to add it there if it isn't. Here is what we did in loopback4-extension-starter (the trick works well with VS Code prettier plugin): 

https://github.com/strongloop/loopback4-extension-starter/blob/3a491d40e7d58ad177484de459050ba127215af7/.prettierignore#L4

```
*.json
```

File: src/controllers/todo.controller.ts
Comment (bajtos) [inline src/controllers/todo.controller.ts]: nitpick: our coding style has no spaces inside/around curly braces

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: Have you considered using data-builder here? See http://loopback.io/doc/en/lb4/Testing-your-application.html#use-test-data-builders

Feel free to make this change later, after this long-living PR is landed.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: I think it's a bad idea to hard-code id numbers in acceptance tests. If we change the persistence config later (e.g. use a different DB that uses UUID instead of numeric ids, or pre-populate database with some initial Todo instances), then we will have to rework all tests like this one.

Here is an assertion that may be better:

```js
const response = await client.post(...)
  // ...
  expect(200);

// all properties from the request were persisted
expect(response.body).to.containEql(payload);

// verify the record was stored
const repo = (await app.get('repositories.todo')) as TodoRepository;
const todoCreated = await TodoRepository.findById(response.body.id);
expect(todoCreated).to.containEql(payload);
```

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: I find these `as Todo` casts ugly. I am proposing to define `givenTodo` to accept any object containing a subset of Todo properties.

```ts
async function givenTodo(todo: Partial<Todo>) {
 // ...
}
```

Ideally, `givenTodo` should fill in any missing properties that are required by the model.

See http://loopback.io/doc/en/lb4/Testing-your-application.html#use-test-data-builders again.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: Is this still necessary? I thought `checklist` was removed?

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: I think this test should verify that the database was updated correctly.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: Ditto - please check that the database was updated as expected.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: Ditto - please check that the database was updated as expected.

File: test/acceptance/application.test.ts
Comment (bajtos) [inline test/acceptance/application.test.ts]: Can we perhaps move this to a new application method?

```ts
export class TodoApplication extends Application {
  // ...
  
  async getTodoRepository() {
    return (await app.get('repositories.todo')) as TodoRepository;
  }
}
```

(It would be great if the repository mixin adding `app.repository(TodoRepository)` method was also adding `app.getRepository<TodoRepository>()`, but that's out of scope of this PR.)

File: test/unit/controllers/todo.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo.controller.test.ts]: When reading this test, I find it confusing why `createTodo` is supposed to return `todoWithId`. It would be better to setup the behaviour of the relevant stub here in this test, even if it means a bit of duplication. I do realise I recommended to move as much of the shared/repeated setup to a "beforeEach" hook before. The difficult part is to strike the right balance between making the test readable by including all relevant bits and not repeating too much stuff.

I also find this style, where the "act" part is on the same line as the first "assert" part, a bit difficult to quickly comprehend.

Here is how I would write this test myself (minus the comments). The name "aTodoWithId" (prefixed with "a") is hinting that any Todo instance can be used here, that there is nothing special about the property values used. It's a convention I found useful in the past, but we don't necessarily have to apply it in LoopBack too.

```ts
it('creates a Todo', async () => {
  // arrange
  create.resolves(aTodoWithId);
  // act
  const result = await controller.createTodo(todo);
  // assert
  expect(result).to.eql(aTodoWithId);
  sinon.assert.calledWith(create, todo);
});
```

This comment applies to other tests too.

File: test/unit/controllers/todo.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo.controller.test.ts]: So what if we add another required property later, and the code verifying required properties will check that new property before the missing "title"? This acceptance test will incorrectly fail in such case.

I think a more robust solution is to use a data builder to create a valid Todo instance data, and then remove only the `title` property.

File: test/unit/controllers/todo.controller.test.ts
Comment (bajtos) [inline test/unit/controllers/todo.controller.test.ts]: This hooks seems to be repeated in every second-level `describe`, can you move it to the first-level describe to avoid repetition?

**Review by bajtos — APPROVED** [review]
The code looks much better now!

There are few more opportunities to improve it, I'll leave it up to you to decide which of them to apply in this pull request and which of them to defer for later.

**Comment by bajtos** [comment]
@kjdelisle it would be great to capture some of my comments inside the tests to explain to readers why the tests are written in the way they are written. Without that, people will not realize what parts of this example are best practices they should follow, and what parts are just details specific to our example app.

File: src/controllers/todo.controller.ts
Comment (kjdelisle) [reply src/controllers/todo.controller.ts]: I'm not sure why Prettier isn't picking up our configuration and formatting it accordingly...

File: test/acceptance/application.test.ts
Comment (kjdelisle) [reply test/acceptance/application.test.ts]: I think I'll make that change as a part of this one, I agree that it's a nicer pattern.

File: test/acceptance/application.test.ts
Comment (kjdelisle) [reply test/acceptance/application.test.ts]: I thought I changed all of those; could be a bad merge.

File: test/acceptance/application.test.ts
Comment (kjdelisle) [reply test/acceptance/application.test.ts]: Merge conflict resolution must've been messed up. :(
That's my fault.

**Pull request merged**
